## Introduction

- Welcome to Data Science with Timmy. 
- Here we are working on a dataset named Flood Prediction. 
- Dataset link: https://www.kaggle.com/competitions/playground-series-s4e5

## Procedure

### Data Understanding
- **Overview**
- Welcome to the 2024 Kaggle Playground Series! We plan to continue in the spirit of previous playgrounds, providing interesting an approachable datasets for our community to practice their machine learning skills, and anticipate a competition each month.

- **Your Goal**: The goal of this competition is to predict the probability of a region flooding based on various factors.
- **Flood detection** refers to the process of identifying, monitoring, and alerting authorities or individuals about the presence or likelihood of flooding in a particular area. It involves the use of various technologies and methods to detect, predict, and mitigate the impacts of floods.

### Let's read the data

In [15]:
# importing some libraries
import pandas as pd #for data manipulation
import numpy as np     #for numerical operations
import matplotlib.pyplot as plt    #for data visualization
import seaborn as sns   #for data visualization
# setting column width
pd.set_option('display.max_columns', None)

There are three files in the zip file.
- train.csv
- test.csv
- sample_submission.csv

In [3]:
# loading the dataset
train = pd.read_csv('train.csv')
test = pd.read_csv('test.csv')
sample = pd.read_csv('sample_submission.csv')

Looking at the top 10 rows in each file

In [4]:
train.head(10)

id  MonsoonIntensity  TopographyDrainage  RiverManagement  Deforestation  \
0   0                 5                   8                5              8   
1   1                 6                   7                4              4   
2   2                 6                   5                6              7   
3   3                 3                   4                6              5   
4   4                 5                   3                2              6   
5   5                 5                   4                1              4   
6   6                 8                   3                1              2   
7   7                 6                   6                5              7   
8   8                 5                   2                8              5   
9   9                 4                   2                3              5   

   Urbanization  ClimateChange  DamsQuality  Siltation  AgriculturalPractices  \
0             6              4            4          3                      3   
1             8              8            3          5                      4   
2             3              7            1          5                      4   
3             4              8            4          7                      6   
4             4              4            3          3                      3   
5             2              4            6          6                      7   
6             3              7            3          4                      6   
7             5              5            3          5                      5   
8             4              5            2          4                      5   
9             8              6            5          5                      7   

   ...  DrainageSystems  CoastalVulnerability  Landslides  Watersheds  \
0  ...                5                     3           3           5   
1  ...                7                     2           0           3   
2  ...                7                     3           7           5   
3  ...                2                     4           7           4   
4  ...                2                     2           6           6   
5  ...                3                     5           5           4   
6  ...                2                     5           6           4   
7  ...                5                     3           5           5   
8  ...                9                     2           7           3   
9  ...                6                     3           3           4   

   DeterioratingInfrastructure  PopulationScore  WetlandLoss  \
0                            4                7            5   
1                            5                3            3   
2                            6                8            2   
3                            4                6            5   
4                            4                1            2   
5                            4                6            8   
6                            5                6            3   
7                            8                6            8   
8                            4                6            4   
9                            4                3            3   

   InadequatePlanning  PoliticalFactors  FloodProbability  
0                   7                 3             0.445  
1                   4                 3             0.450  
2                   3                 3             0.530  
3                   7                 5             0.535  
4                   3                 5             0.415  
5                   3                 2             0.440  
6                   4                 6             0.460  
7                   5                 6             0.595  
8                   5                 5             0.505  
9                   5                 6             0.455  

[10 rows x 22 columns]

In [5]:
test.head(10)

id  MonsoonIntensity  TopographyDrainage  RiverManagement  \
0  1117957                 4                   6                3   
1  1117958                 4                   4                2   
2  1117959                 1                   3                6   
3  1117960                 2                   4                4   
4  1117961                 6                   3                2   
5  1117962                 3                   4                8   
6  1117963                 5                   5                6   
7  1117964                 6                   9                6   
8  1117965                 6                   5                2   
9  1117966                 4                   8                8   

   Deforestation  Urbanization  ClimateChange  DamsQuality  Siltation  \
0              5             6              7            8          7   
1              9             5              5            4          7   
2              5             7              2            4          6   
3              6             4              5            4          3   
4              4             6              4            5          5   
5              4             6              5            2          5   
6              8             6              5            3          6   
7              5             5              5            3          8   
8              4             6              4            5          8   
9              4             7              7            4          1   

   AgriculturalPractices  ...  IneffectiveDisasterPreparedness  \
0                      8  ...                                8   
1                      5  ...                                2   
2                      4  ...                                7   
3                      4  ...                                7   
4                      3  ...                                4   
5                      9  ...                                6   
6                      9  ...                                6   
7                     10  ...                                3   
8                      4  ...                                4   
9                      3  ...                                7   

   DrainageSystems  CoastalVulnerability  Landslides  Watersheds  \
0                5                     7           5           6   
1                4                     7           4           5   
2                9                     2           5           5   
3                8                     4           6           7   
4                3                     2           6           4   
5                4                     3           7           1   
6                4                     2           5           5   
7                5                     6           1           2   
8                2                     6           6           7   
9                8                     4           4           7   

   DeterioratingInfrastructure  PopulationScore  WetlandLoss  \
0                            3                6            4   
1                            1                7            4   
2                            2                3            6   
3                            6                4            2   
4                            6                8            4   
5                            6                6            3   
6                            5                4            6   
7                            5                3            6   
8                            3                3            6   
9                           10                3            6   

   InadequatePlanning  PoliticalFactors  
0                   4                 5  
1                   4                 3  
2                   8                 3  
3                   4                 4  
4                   5                 5  
5          

In [6]:
sample.head(10)

id  FloodProbability
0  1117957               0.5
1  1117958               0.5
2  1117959               0.5
3  1117960               0.5
4  1117961               0.5
5  1117962               0.5
6  1117963               0.5
7  1117964               0.5
8  1117965               0.5
9  1117966               0.5

What next? 

## Data Exploration 
- We want to see more about the data since we only just looked at it. 

In [7]:
# shape of our data
print(train.shape)
print(test.shape)

(1117957, 22)
(745305, 21)


Wow!!!!!
- We have quite a large data
- Training set has over 1 million rows and 22 columns
- Testing set has 700k plus rows and 21 columns.

- Why the difference in columns though? 
- That's because we have a target variable in our train set which is absent in the test set.

In [8]:
# checking for the columns in the dataset
train.columns

Index(['id', 'MonsoonIntensity', 'TopographyDrainage', 'RiverManagement',
       'Deforestation', 'Urbanization', 'ClimateChange', 'DamsQuality',
       'Siltation', 'AgriculturalPractices', 'Encroachments',
       'IneffectiveDisasterPreparedness', 'DrainageSystems',
       'CoastalVulnerability', 'Landslides', 'Watersheds',
       'DeterioratingInfrastructure', 'PopulationScore', 'WetlandLoss',
       'InadequatePlanning', 'PoliticalFactors', 'FloodProbability'],
      dtype='object')

In [9]:
test.columns

Index(['id', 'MonsoonIntensity', 'TopographyDrainage', 'RiverManagement',
       'Deforestation', 'Urbanization', 'ClimateChange', 'DamsQuality',
       'Siltation', 'AgriculturalPractices', 'Encroachments',
       'IneffectiveDisasterPreparedness', 'DrainageSystems',
       'CoastalVulnerability', 'Landslides', 'Watersheds',
       'DeterioratingInfrastructure', 'PopulationScore', 'WetlandLoss',
       'InadequatePlanning', 'PoliticalFactors'],
      dtype='object')

What's the difference? 

Here is the difference:
- train set has Flood Probability
- test set does not have Flood Probability

What have I seen now? 
- The Flood Probability in my train set is the TARGET VARIABLE!!!!!!.
- In short We are to Predict the Flood Probability of the data points in the test data.

### Are we done with EDA? NO, Let's GOOO!!!!!!

- We are yet to fully understand our data of we don't know the features of our data very well.
- Let's do that!!!!!!

In [10]:
# looking at out features again
train.columns

Index(['id', 'MonsoonIntensity', 'TopographyDrainage', 'RiverManagement',
       'Deforestation', 'Urbanization', 'ClimateChange', 'DamsQuality',
       'Siltation', 'AgriculturalPractices', 'Encroachments',
       'IneffectiveDisasterPreparedness', 'DrainageSystems',
       'CoastalVulnerability', 'Landslides', 'Watersheds',
       'DeterioratingInfrastructure', 'PopulationScore', 'WetlandLoss',
       'InadequatePlanning', 'PoliticalFactors', 'FloodProbability'],
      dtype='object')

- Making Research on them. 
- Ask ChatGPT questions, relate it to the problem statement for better understanding. 

## Research Results

- Monsoon Intensity: This feature likely measures the strength or intensity of monsoon rains in the region. Monsoons are seasonal winds that bring heavy rainfall to certain parts of the world, often causing floods.

- Topography Drainage: This feature could represent the natural drainage characteristics of the area's topography, indicating how well water flows away from the land surface. Areas with poor drainage might be more prone to flooding.

- River Management: This feature might describe the effectiveness of river management practices in the region, such as dam construction, levees, and floodplain zoning.

- Deforestation: This feature likely measures the extent of deforestation in the area. Deforestation can increase surface runoff and soil erosion, contributing to flood risk.

- Urbanization: This feature may indicate the degree of urban development in the region. Urbanization often leads to increased impervious surfaces like pavement and buildings, which can exacerbate flooding by reducing natural infiltration and increasing runoff.

- Climate Change: This feature could represent the influence of climate change on flood risk factors, such as altered precipitation patterns, sea level rise, and increased frequency or intensity of extreme weather events.

- Dams Quality: This feature likely assesses the condition and effectiveness of dams in the region for flood control and water management purposes.

- Siltation: Siltation refers to the accumulation of sediment in rivers, reservoirs, and other water bodies. Excessive siltation can reduce water storage capacity and increase flood risk.

- Agricultural Practices: This feature may describe agricultural activities and land use practices that can impact flood risk, such as soil management, irrigation practices, and land drainage.

- Encroachments: This feature likely indicates the extent of encroachment or unauthorized occupation of floodplains and other natural waterways, which can exacerbate flood impacts.

- Disaster Preparedness: This feature could represent the level of preparedness and response capabilities of local authorities and communities in dealing with flood disasters.

- Drainage Systems: This feature may describe the condition and effectiveness of drainage infrastructure, including stormwater drains, culverts, and canals.

- Coastal Vulnerability: This feature likely assesses the susceptibility of coastal areas to flooding and storm surges, considering factors such as sea level rise, coastal erosion, and tidal patterns.

- Landslides: This feature may indicate the risk of landslides in the region, which can be triggered by heavy rainfall and contribute to downstream flooding.

- Watersheds: Watersheds are areas of land where all the water drains into a common outlet, such as a river or lake. This feature may represent the characteristics of watersheds in the region and their influence on flooding.

- Deteriorating Infrastructure: This feature likely describes the condition of critical infrastructure systems (e.g., roads, bridges, levees) that play a role in flood mitigation and response.

- Population Score: This feature may indicate population density or other demographic factors that can influence flood vulnerability and the potential impact on human lives and property.

- Wetland Loss: This feature likely measures the extent of wetland loss in the region, which can reduce natural flood storage capacity and increase flood risk.

- Inadequate Planning: This feature could represent deficiencies in land use planning, zoning regulations, and development policies that contribute to increased flood vulnerability.

- Political Factors: This feature may encompass political considerations and governance issues that affect flood risk management, such as corruption, regulatory enforcement, and allocation of resources.

- Flood Probability: This is likely the target variable you're trying to predict, representing the likelihood or probability of flooding occurring in a given area based on the combination of all the aforementioned factors.

## Is it Much? 

- Doesn't matter........
- What matters is that you understand your data. 


## Now what next? 

- MORE EDA,  we are just getting started.
- We just understood our data. 

In [11]:
# checking for the data types
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1117957 entries, 0 to 1117956
Data columns (total 22 columns):
 #   Column                           Non-Null Count    Dtype  
---  ------                           --------------    -----  
 0   id                               1117957 non-null  int64  
 1   MonsoonIntensity                 1117957 non-null  int64  
 2   TopographyDrainage               1117957 non-null  int64  
 3   RiverManagement                  1117957 non-null  int64  
 4   Deforestation                    1117957 non-null  int64  
 5   Urbanization                     1117957 non-null  int64  
 6   ClimateChange                    1117957 non-null  int64  
 7   DamsQuality                      1117957 non-null  int64  
 8   Siltation                        1117957 non-null  int64  
 9   AgriculturalPractices            1117957 non-null  int64  
 10  Encroachments                    1117957 non-null  int64  
 11  IneffectiveDisasterPreparedness  1117957 non-null 

Our EDA will be focused on the train set initially. 

- This EDA shows that we only have numerical features in out Data....

In [12]:
# checking for unique values in the dataset
train.nunique()

id                                 1117957
MonsoonIntensity                        17
TopographyDrainage                      19
RiverManagement                         17
Deforestation                           18
Urbanization                            18
ClimateChange                           18
DamsQuality                             17
Siltation                               17
AgriculturalPractices                   17
Encroachments                           19
IneffectiveDisasterPreparedness         17
DrainageSystems                         18
CoastalVulnerability                    18
Landslides                              17
Watersheds                              17
DeterioratingInfrastructure             18
PopulationScore                         19
WetlandLoss                             20
InadequatePlanning                      17
PoliticalFactors                        17
FloodProbability                        83
dtype: int64

In [13]:
# checking the data again.

train.head()

id  MonsoonIntensity  TopographyDrainage  RiverManagement  Deforestation  \
0   0                 5                   8                5              8   
1   1                 6                   7                4              4   
2   2                 6                   5                6              7   
3   3                 3                   4                6              5   
4   4                 5                   3                2              6   

   Urbanization  ClimateChange  DamsQuality  Siltation  AgriculturalPractices  \
0             6              4            4          3                      3   
1             8              8            3          5                      4   
2             3              7            1          5                      4   
3             4              8            4          7                      6   
4             4              4            3          3                      3   

   ...  DrainageSystems  CoastalVulnerability  Landslides  Watersheds  \
0  ...                5                     3           3           5   
1  ...                7                     2           0           3   
2  ...                7                     3           7           5   
3  ...                2                     4           7           4   
4  ...                2                     2           6           6   

   DeterioratingInfrastructure  PopulationScore  WetlandLoss  \
0                            4                7            5   
1                            5                3            3   
2                            6                8            2   
3                            4                6            5   
4                            4                1            2   

   InadequatePlanning  PoliticalFactors  FloodProbability  
0                   7                 3             0.445  
1                   4                 3             0.450  
2                   3                 3             0.530  
3                   7                 5             0.535  
4                   3                 5             0.415  

[5 rows x 22 columns]

In [16]:
#  A lil preview of the data by performing the describe function
train.describe()

id  MonsoonIntensity  TopographyDrainage  RiverManagement  \
count  1.117957e+06      1.117957e+06        1.117957e+06     1.117957e+06   
mean   5.589780e+05      4.921450e+00        4.926671e+00     4.955322e+00   
std    3.227265e+05      2.056387e+00        2.093879e+00     2.072186e+00   
min    0.000000e+00      0.000000e+00        0.000000e+00     0.000000e+00   
25%    2.794890e+05      3.000000e+00        3.000000e+00     4.000000e+00   
50%    5.589780e+05      5.000000e+00        5.000000e+00     5.000000e+00   
75%    8.384670e+05      6.000000e+00        6.000000e+00     6.000000e+00   
max    1.117956e+06      1.600000e+01        1.800000e+01     1.600000e+01   

       Deforestation  Urbanization  ClimateChange   DamsQuality     Siltation  \
count   1.117957e+06  1.117957e+06   1.117957e+06  1.117957e+06  1.117957e+06   
mean    4.942240e+00  4.942517e+00   4.934093e+00  4.955878e+00  4.927791e+00   
std     2.051689e+00  2.083391e+00   2.057742e+00  2.083063e+00  2.065992e+00   
min     0.000000e+00  0.000000e+00   0.000000e+00  0.000000e+00  0.000000e+00   
25%     4.000000e+00  3.000000e+00   3.000000e+00  4.000000e+00  3.000000e+00   
50%     5.000000e+00  5.000000e+00   5.000000e+00  5.000000e+00  5.000000e+00   
75%     6.000000e+00  6.000000e+00   6.000000e+00  6.000000e+00  6.000000e+00   
max     1.700000e+01  1.700000e+01   1.700000e+01  1.600000e+01  1.600000e+01   

       AgriculturalPractices  Encroachments  IneffectiveDisasterPreparedness  \
count           1.117957e+06   1.117957e+06                     1.117957e+06   
mean            4.942619e+00   4.949230e+00                     4.945239e+00   
std             2.068545e+00   2.083324e+00                     2.078141e+00   
min             0.000000e+00   0.000000e+00                     0.000000e+00   
25%             3.000000e+00   4.000000e+00                     3.000000e+00   
50%             5.000000e+00   5.000000e+00                     5.000000e+00   
75%             6.000000e+00   6.000000e+00                     6.000000e+00   
max             1.600000e+01   1.800000e+01                     1.600000e+01   

       DrainageSystems  CoastalVulnerability    Landslides    Watersheds  \
count     1.117957e+06          1.117957e+06  1.117957e+06  1.117957e+06   
mean      4.946893e+00          4.953999e+00  4.931376e+00  4.929032e+00   
std       2.072333e+00          2.088899e+00  2.078287e+00  2.082395e+00   
min       0.000000e+00          0.000000e+00  0.000000e+00  0.000000e+00   
25%       4.000000e+00          3.000000e+00  3.000000e+00  3.000000e+00   
50%       5.000000e+00          5.000000e+00  5.000000e+00  5.000000e+00   
75%       6.000000e+00          6.000000e+00  6.000000e+00  6.000000e+00   
max       1.700000e+01          1.700000e+01  1.600000e+01  1.600000e+01   

       DeterioratingInfrastructure  PopulationScore   WetlandLoss  \
count                 1.117957e+06     1.117957e+06  1.117957e+06   
mean                  4.925907e+00     4.927520e+00  4.950859e+00   
std                   2.064813e+00     2.074176e+00  2.068696e+00   
min                   0.000000e+00     0.000000e+00  0.000000e+00   
25%                   3.000000e+00     3.000000e+00  4.000000e+00   
50%                   5.000000e+00     5.000000e+00  5.000000e+00   
75%                   6.000000e+00     6.000000e+00  6.000000e+00   
max                   1.700000e+01     1.800000e+01  1.900000e+01   

       InadequatePlanning  PoliticalFactors  FloodProbability  
count        1.117957e+06      1.117957e+06      1.117957e+06  
mean         4.940587e+00      4.939004e+00      5.044803e-01  
std          2.081123e+00      2.090350e+00      5.102610e-02  
min          0.000000e+00      0.000000e+00      2.850000e-01  
25%          3.000000e+00      3.000000e+00      4.700000e-01  
50%          5.000000e+00      5.000000e+00      5.050000e-01  
75%          6.000000e+00      6.000000e+00      5.400000e-01  
max          1.600000e+01      1.600000e+01 

In [17]:
# displaying the min and max values of the dataset
train.min()

id                                 0.000
MonsoonIntensity                   0.000
TopographyDrainage                 0.000
RiverManagement                    0.000
Deforestation                      0.000
Urbanization                       0.000
ClimateChange                      0.000
DamsQuality                        0.000
Siltation                          0.000
AgriculturalPractices              0.000
Encroachments                      0.000
IneffectiveDisasterPreparedness    0.000
DrainageSystems                    0.000
CoastalVulnerability               0.000
Landslides                         0.000
Watersheds                         0.000
DeterioratingInfrastructure        0.000
PopulationScore                    0.000
WetlandLoss                        0.000
InadequatePlanning                 0.000
PoliticalFactors                   0.000
FloodProbability                   0.285
dtype: float64

In [18]:
train.max()

id                                 1117956.000
MonsoonIntensity                        16.000
TopographyDrainage                      18.000
RiverManagement                         16.000
Deforestation                           17.000
Urbanization                            17.000
ClimateChange                           17.000
DamsQuality                             16.000
Siltation                               16.000
AgriculturalPractices                   16.000
Encroachments                           18.000
IneffectiveDisasterPreparedness         16.000
DrainageSystems                         17.000
CoastalVulnerability                    17.000
Landslides                              16.000
Watersheds                              16.000
DeterioratingInfrastructure             17.000
PopulationScore                         18.000
WetlandLoss                             19.000
InadequatePlanning                      16.000
PoliticalFactors                        16.000
FloodProbabil

## Let's make some inference on what we've seen in our data.

- 1. The minimum values for the features in the data are zero (target variable excluded)
- 2. The maximum values are between 16 and 19.
- 3. We only have numerical features in the data, no categorical.
- 4. Based on the data description, there's a high chance that we have outliers in our data. (we'll see more about this)
- 5. The target variable (Flood Probability) values are between 0.285 - 0.725
- 6. We have 20 features (ignoring the id column). 